build/utils-node.js

```diff
@@ -10,179 +10,540 @@
+    } else {
+      webapp.appStatus = utils.getAppStatus(webapp.manifest.type);
+    }
+
+    // Some webapps control their own build
+    webapp.buildDirectoryFile = this.getFile(config.STAGE_DIR,
+      webapp.sourceDirectoryName);
+    webapp.buildManifestFile = this.getFile(webapp.buildDirectoryFile.path,
+      'manifest.webapp');
+
+    // Generate the webapp folder name in the profile. Only if it's privileged
+    // and it has an origin in its manifest file it'll be able to specify a
+    // custom folder name. Otherwise, generate an UUID to use as folder name.
+    var webappTargetDirName;
+    if (this.isExternalApp(webapp)) {
+      var type = webapp.appStatus;
+      var isPackaged = false;
+      if (webapp.pckManifest) {
+        isPackaged = true;
+        if (webapp.metaData.origin) {
+          throw new Error('External webapp `' + webapp.sourceDirectoryName +
+            '` can not have origin in metadata because is packaged');
+        }
+      }
+      if (type === 2 && isPackaged && webapp.pckManifest.origin) {
+        webappTargetDirName = this.getNewURI(webapp.pckManifest.origin).host;
+      } else {
+        // uuid is used for webapp directory name, save it for further usage
+        let mapping = this.getUUIDMapping(config);
+        var uuid = mapping[webapp.sourceDirectoryName] || nodeUUID.v4();
+        mapping[webapp.sourceDirectoryName] = webappTargetDirName = uuid;
+      }
+    } else {
+      webappTargetDirName = webapp.domain;
+    }
+    webapp.profileDirectoryFile = this.getFile(config.PROFILE_DIR, 'webapps',
+      webappTargetDirName);
+
+    return webapp;
+  },
+
+  getNewURI: function(uri) {
+    return url.parse(uri);
+  },
+
+  isExternalApp: function(webapp) {
+    if (webapp.metaData && webapp.metaData.external === undefined) {
+      throw new Error('"external" property in metadata.json is required ' +
+        'since Firefox OS 2.1, please add it into metadata.json and update ' +
+        'preload.py if you use this script to perload your apps. If you ' +
+        'created metadata.json for non-external apps, please set "external" ' +
+        'to false. metadata.json is in ' + webapp.sourceDirectoryFile.path);
+    }
+    if (!webapp.metaData || webapp.metaData.external === false) {
+      return false;
+    } else {
+      return true;
+    }
+  },
+
+  processEvents: function() {
+
+  },
+
+  writeContent :function(file, content) {
+    fs.writeFileSync(file.path, content);
+  },
+
+  createZip: function() {
+    return new JSZip();
+  },
+
+  getCompression: function(type) {
+    switch(type) {
+      case 'none':
+        return 'STORE';
+      case 'best':
+        return 'DEFLATE';
+    }
+  },
+
+  addFileToZip: function(zip, zipPath, pathInZip, file, compression) {
+    if (!file.exists()) {
+      return;
+    }
+
+    zip.file(pathInZip, fs.readFileSync(file.path), {
+      compression: compression || 'DEFLATE'
+    });
+  },
```

Comment:
addEntryContentWithTime() should be rewritten because it lots of differences from JSZip. I propose we should have a addFileToZip() tentatively.